source/MaterialXRuntime/Private/PvtMessage.cpp

```diff
@@ -52,6 +52,30 @@ RtCallbackId PvtMessageHandler::addSetAttributeCallback(RtSetAttributeCallbackFu
     return _callbackIdCounter++;
 }
 
+RtCallbackId PvtMessageHandler::addSetMetadataCallback(RtSetMetadataCallbackFunc callback, void* userData)
+{
+    PvtSetMetadataObserver observer = PvtSetMetadataObserver(callback, userData);
+    _setMetadataObservers[_callbackIdCounter] = observer;
+    _callbackIdToType[_callbackIdCounter] = observer.type;
+    return _callbackIdCounter++;
+}
+
+RtCallbackId PvtMessageHandler::addCreateMetadataCallback(RtCreateMetadataCallbackFunc callback, void* userData)
```

Comment:
I'm curious about this observation "model" instead of have an observer that can receieve different states, you need 1 custom observer per state type change ? Guess this works but it could end up being a fair number of observers to have to maintain. Guess in general there's could be a large number of observers required to cover all the different changes that can occur and if you miss any then something will slip through unobserved.

Reply:
I agree with your point, but I didn't implement this, I'm just following the pattern. Maybe @niklasharrysson has some thoughts?

Reply:
The reason there are separate observers per change type is runtime performance. If an observer is only interested in a single type of change we don't want to send _all_ changes to that observer. Even if the receiving observer just ignores the message the fact we are sending it takes a bit of processing.

But note that these are just implementation details in the private API. A user of the public API never sees the observer classes. As an API user the only thing you need to do is to start subscribing to a change type on the RtMessage class:
https://github.com/autodesk-forks/MaterialX/blob/adsk_contrib/dev/source/MaterialXRuntime/RtMessage.h#L55
And your callback will be called when this type of change happens.